edit-event/DurationPicker.tsx

```diff
@@ -1,9 +1,399 @@
+        marginLeft: withTiFDefaultSpring(
+          !isSliding.value ? INNER_TRACK_GAP : 0
+        ),
+        borderBottomLeftRadius: withTiFDefaultSpring(
+          isSliding.value ? INNER_TRACK_GAP * 2 : INNER_TRACK_GAP
+        ),
+        borderTopLeftRadius: withTiFDefaultSpring(
+          isSliding.value ? INNER_TRACK_GAP * 2 : INNER_TRACK_GAP
+        )
+      }))
+    ]}
+  />
+)
+
+type DurationsRowProps = {
+  durations: number[]
+  height: number
+  onDurationLayout: (duration: number, layout: LayoutRectangle) => void
+  onDurationChange: (duration: number) => void
+}
+
+const DurationsRowView = ({
+  durations,
+  height,
+  onDurationLayout,
+  onDurationChange
+}: DurationsRowProps) => (
+  <View style={[styles.presetsRow, { height }]}>
+    {durations.map((duration) => (
+      <Pressable
+        key={`d-${duration}`}
+        style={styles.presetItem}
+        onPress={() => onDurationChange(duration)}
+        onLayout={(e) => onDurationLayout(duration, e.nativeEvent.layout)}
+      >
+        <BodyText
+          maxFontSizeMultiplier={FontScaleFactors.large}
+          style={styles.presetText}
+        >
+          {formatEventDurationPreset(duration)}
+        </BodyText>
+      </Pressable>
+    ))}
+  </View>
+)
+
+type SliderKnobProps = {
+  durationAtom: PrimitiveAtom<number>
+  sliderPosition: SharedValue<number>
+  isSliding: SharedValue<boolean>
+  pickerState: EditEventDurationPickerState
+  height: number
+}
+
+const SliderKnobView = ({
+  durationAtom,
+  sliderPosition,
+  isSliding,
+  pickerState,
+  height
+}: SliderKnobProps) => {
+  const [duration, setDuration] = useAtom(durationAtom)
+  const stateEntries = pickerStateEntries(pickerState)
+  const selectedDimensions = layoutForDuration(stateEntries, duration)
+  const animateToPosition = useEffectEvent((dimensions?: LayoutRectangle) => {
+    if (!isSliding.value) {
+      sliderPosition.value = withTiFDefaultSpring(dimensions?.x ?? 0)
+    }
+  })
+  const previousTranslation = useSharedValue(0)
+  useEffect(
+    () => animateToPosition(selectedDimensions),
+    [selectedDimensions, animateToPosition]
+  )
+  const panGestureEndBound = endBound(pickerState)
+  const panGesture = Gesture.Pan()
+    .onUpdate((event) => {
+      sliderPosition.value = clamp(
+        -INNER_TRACK_GAP,
+        panGestureEndBound,
+        sliderPosition.value + (event.translationX - previousTranslation.value)
+      )
+      previousTranslation.value = event.translationX
+      const nextDuration = durationAtPosition(
+        stateEntries,
+        sliderPosition.value
+      )
+      if (nextDuration !== duration) runOnJS(setDuration)(nextDuration)
+    })
+    .onFinalize(() => {
+      const nextPosition =
+        sliderPosition.value < 0
+          ? 0
+          : layoutForDuration(stateEntries, duration)?.x ?? 0
+      sliderPosition.value = withTiFDefaultSpring(nextPosition)
+      runOnJS(setDuration)(durationAtPosition(stateEntries, nextPosition))
```

Comment:
How is runOnJS being used here? From what I could tell from research, it calls `setDuration` asynchronously, but I'm not quite sure how this would work.
From what I assume, it's to update the current duration (`duration`) if it doesn't match the `nextDuration` given by the picker itself, on `onUpdate.` For `onFinalize`, I'm not sure.

Reply:
Reanimated doesn't let you call any given function inside the code for the gesture handler since it runs on a separate thread controlled by native code. JavaScript is a single threaded language, so react native runs the JS on a dedicated thread. This creates a problem for multithreaded code like in this case.

You'll notice that some functions have the string `"worklet"` at the start. This string tells reanimated that this function can be called on the native thread, and the reanimated babel plugin will do the work of making that compatible. `setDuration` is a function that is owned by jotai, so we cannot mark it as `"worklet"`. Therefore, we can get around this with `runOnJs`, which will run `setDuration` on react native's JS thread making everyone happy.